### PR TITLE
Run Percy if a package has a dependency within origami which has changed comapred to the main branch

### PR DESCRIPTION
**Pull request by JakeChampion**
This figures out which components across the whole monorepo are affected by a change.
E.G. If o-spacing has changed, this will run n-notifications Percy even though n-notification does not directly depend on o-spacing. It runs because n-notification depends on o-typography which depends on o-spacing

### CODE REVIEW

File: components/o-fonts/src/scss/_variables.scss [hunk 1]
```diff
@@ -6,7 +6,7 @@ $o-fonts-is-silent: true !default;
 /// Path to default font files.
 ///
 /// @type String
-$o-fonts-path: 'https://www.ft.com/__origami/service/build/v3/font?version=1.12' !default;
+$o-fonts-path: 'https://www.ft.com/__origami/service/build/v3/font?version=^1.12' !default;
```
Comment (JakeChampion) [inline components/o-fonts/src/scss/_variables.scss:9]: @Financial-Times/origami-core this was added to test out the percy configuration, all the components which directly depend on o-fonts as either a dependency, dev-dependency or peer-dependency would have had their percy project run

File: scripts/percy.js [hunk 2]
```diff
@@ -1,29 +1,75 @@
...
+					for (let cwd of await workspaces.paths()) {
+						let pkg = await readPackage({cwd})
+						const packageDependencies = new Set(Object.keys(pkg.dependencies||[]).concat(Object.keys(pkg.devDependencies||[])).concat(Object.keys(pkg.peerDependencies||[])))
+						for (const dependency of packageDependencies) {
+							if (changedPackages.has(dependency)) {
+								changedPackages.add(pkg.name)
+							}
+						}
+					}
+				}
```
Comment (JakeChampion) [inline scripts/percy.js:59]: This section makes this work for any number of nested origami dependencies

File: scripts/percy.js [hunk 3]
```diff
@@ -17,13 +23,53 @@ async function shouldPercyRun() {
...
 			if (message.startsWith('fix:') || message.startsWith('fix!:') || message.startsWith('feat:') || message.startsWith('feat!:')) {
-				core.notice('A commit in the pull-request is a `fix` or `feat` commit - which means some user facing changes have been made. This means we need to run Percy on the pull-request.')
-				return true;
+				$.verbose = false
+				let {stdout: changedFiles} = await $`git diff --name-only origin/${baseRef}...origin/${headRef}`;
+				$.verbose = true
+				const changedPackages = new Set;
+				for (const file of changedFiles.split('\n')) {
+					if (file) {
+						const a = file.split('/');
```
Comment (JakeChampion) [inline scripts/percy.js:37]: I need to come up with a more understandable variable name here

File: package.json [hunk 4]
```diff
@@ -80,7 +80,7 @@
     "@actions/exec": "^1.0.4",
     "@actions/github": "^4.0.0",
     "@actions/io": "^1.0.2",
-    "@percy/cli": "1.0.0-beta.67",
```
Comment (notlee) [inline package.json:83]: 🎉

File: scripts/percy.js (hunk 2 above)
Comment (notlee) [reply scripts/percy.js:61]: thanks for expanding, I had to read a couple of times to understand, would be super nice to get this as a comment in the code or break out into a named function

**Review by notlee — APPROVED** [review]
Couple of non-blocking suggestions :)

File: scripts/percy.js [hunk 5]
```diff
@@ -17,13 +23,58 @@ async function shouldPercyRun() {
...
+				}
+				let pkg = await readPackage({cwd: workspace})
+				if (changedPackages.has(pkg.name)) {
+					core.notice('A commit in the pull-request is a `fix` or `feat` commit - which means some user facing changes have been made. Those user facing changes are in this package. This means we need to run Percy on the pull-request.')
+					return true;
+				}
+				const packageDependencies = new Set(Object.keys(pkg.dependencies||[]).concat(Object.keys(pkg.devDependencies||[])).concat(Object.keys(pkg.peerDependencies||[])))
+				for (const dependency of packageDependencies) {
+					if (changedPackages.has(dependency)) {
+						core.notice('A commit in the pull-request is a `fix` or `feat` commit - which means some user facing changes have been made. Those user facing changes are in a dependency of this package. This means we need to run Percy on the pull-request.')
```
Comment (JakeChampion) [inline scripts/percy.js:73]:
```suggestion
						core.notice('A commit in the pull-request is a `fix` or `feat` commit - which means some user facing changes have been made. Those user facing changes are in a dependency of this package. This means we need to run Percy on the pull-request.')
						core.notice(`The dependencies for this package (${pkg.name}) are: ${packageDependencies.join(', ')}`)
```

File: scripts/percy.js [hunk 6]
```diff
@@ -17,13 +23,55 @@ async function shouldPercyRun() {
...
```
[the remark this answers is not in the document]
Comment (notlee) [reply scripts/percy.js:73]: Oops https://github.com/Financial-Times/origami/pull/703/files/7f280c3b948bff71cf8c95e6f4f5c1f453e28c46..60c78eaad8d0b4352d12d1e2d4a620959c8a2881#r847411997

File: scripts/percy.js [hunk 7]
```diff
@@ -71,6 +71,7 @@ async function shouldPercyRun() {
 				for (const dependency of packageDependencies) {
 					if (changedPackages.has(dependency)) {
 						core.notice('A commit in the pull-request is a `fix` or `feat` commit - which means some user facing changes have been made. Those user facing changes are in a dependency of this package. This means we need to run Percy on the pull-request.')
+						core.notice(`The dependencies for this package (${pkg.name}) are: ${packageDependencies.join(', ')}`)
```
[the remark this answers is not in the document]
Comment (JakeChampion) [reply scripts/percy.js:74]: That can be a subsequent PR :D

**Pull request merged**